qiskit/dagcircuit/dagcircuit.py

```diff
@@ -1025,15 +1025,12 @@ def layers(self):
         except StopIteration:
             return
 
-        def add_nodes_from(layer, nodes):
-            """ Convert DAGNodes into a format that can be added to a
-             multigraph and then add to graph"""
-            layer._multi_graph.add_nodes_from(nodes)
-
         for graph_layer in graph_layers:
 
             # Get the op nodes from the layer, removing any input and output nodes.
+            # Sort to make sure they are in the order they were added to the original DAG
             op_nodes = [node for node in graph_layer if node.type == "op"]
+            op_nodes.sort(key=lambda nd: nd._node_id)
```

Comment:
Ideally, a topological sort is what is needed. However we are getting the nodes from a list of nodes, not a graph, so the idea of topology has been lost and the `._node_id` is the only part which could imply the ordering. The lists are created by `multigraph_layers()` which creates the layers based on the successors of nodes, also using lists and so without implying the topology. This sort could be moved to that method instead of having it here, but some form of sort is needed. It was the case that nodes were appearing in an order different than the one they were declared in, the example that pulled it up was the CX gates [here](https://github.com/Qiskit/qiskit-terra/blob/ae1921a41a481ae90b12ea7c494c19fdbf4f70fe/test/python/visualization/test_circuit_text_drawer.py#L643).

Reply:
Ah, I think I see. If I understand how we define a layer, it's entirely okay to not have any topological ordering, since nodes within a layer should not have a dependence between them on the DAG. If they did, they would need to be in two separate layers.

In this case, the need for sorting by `_node_id` is because we have drawing tools that rely on `_node_id` to infer order of node creation and so we need that ordering to be preserved by `layers`. That's fine, just potentially confusing so we should document it.

Reply:
Yes I agree, the nodes inside a layer shouldn't have any ordering. 

The code that relies on it is using the `dag.layers()` [here](https://github.com/Qiskit/qiskit-terra/blob/ae1921a41a481ae90b12ea7c494c19fdbf4f70fe/qiskit/visualization/utils.py#L104) to generate the layers to be drawn (adding in the justification). Although this is acting on a DAG, the DAG is the one that is created by `layers()` and so if the order has not been maintained in this method, it cannot be used by the drawing tools.

I will update the comments to make this clearer.